blocks/blockstore/blockstore.go

```diff
@@ -89,37 +90,62 @@ func (bs *blockstore) Get(k key.Key) (*blocks.Block, error) {
 	return blocks.NewBlockWithHash(bdata, mh.Multihash(k))
 }
 
-func (bs *blockstore) Put(block *blocks.Block) error {
+func (bs *blockstore) Put(block blocks.Block) error {
 	k := block.Key().DsKey()
 
-	// Has is cheaper than Put, so see if we already have it
-	exists, err := bs.datastore.Has(k)
-	if err == nil && exists {
-		return nil // already stored.
+	data := bs.prepareBlock(k, block)
+	if data == nil {
+		return nil
 	}
-	return bs.datastore.Put(k, block.Data)
+	return bs.datastore.Put(k, data)
 }
 
-func (bs *blockstore) PutMany(blocks []*blocks.Block) error {
+func (bs *blockstore) PutMany(blocks []blocks.Block) error {
 	t, err := bs.datastore.Batch()
 	if err != nil {
 		return err
 	}
 	for _, b := range blocks {
 		k := b.Key().DsKey()
-		exists, err := bs.datastore.Has(k)
-		if err == nil && exists {
+		data := bs.prepareBlock(k, b)
+		if data == nil {
 			continue
 		}
-
-		err = t.Put(k, b.Data)
+		err = t.Put(k, data)
 		if err != nil {
 			return err
 		}
 	}
 	return t.Commit()
 }
 
+func (bs *blockstore) prepareBlock(k ds.Key, block blocks.Block) interface{} {
+	if fsBlock, ok := block.(*blocks.FilestoreBlock); !ok {
+		// Has is cheaper than Put, so see if we already have it
+		exists, err := bs.datastore.Has(k)
```

Comment:
@whyrusleeping if a block is already in the normal datastore and a user adds the file with "--nocopy" I want the file to be added to the filestore so that the block in the normal datastore can eventually be removed to save space.  That is why when we have a FilestoreBlock I skip the check Has() check.  I hope this makes sense.